### ISSUE TITLE
Lua API questions

### ISSUE THREAD

**Issue by gamax92** [SourMesen/Mesen#259]
Interested in porting some old FCEUX scripts to the Lua API available in Mesen, but running into a couple of issues.

Wondering why the savestate function must be called only in a start frame callback, the script in question creates savestates before certain instructions are executed, experimenting with modifying their behavior and reverting if something goes wrong. There's an `emu.getState/setState` function but the docs for this say they can't modify apu or cart state, so it would be a partial restore at best.

Also, is there a way to get general keyboard input/extra bindings specifically for triggering things in the Lua script, I've only seen a function that reads the state of the NES controller.

**Comment by SourMesen** [comment]
The savestate function is restricted to the start frame callback for a couple of reasons.  A minor one is that that's also where Mesen takes its regular savestates (so it was the simplest way of ensuring that they work properly without testing a lot of edge cases).

The main reason though, is that most other callbacks can happen in the middle of executing a CPU instruction (e.g the write callback happens in the middle of the cpu instruction, just before the write occurs, which allows you to alter the written value via the Lua script, for example) and savestates can't be taken in the middle of instructions, only in-between instructions. This is a bit of a bigger problem. I would need to implement some way of delaying the save state until the current instruction is over.  All the solutions I can think of at the moment aren't too Lua-developer-friendly, though - I'll see if I can find a decent solution to this.

For the keyboard, there's no way to check specific keys at the moment. (You can check the mouse's buttons, though).  It would be easy to add something like `emu.isKeyPressed("Space")`, though.  It would also work for gamepad buttons `emu.isKeyPressed("Pad1 A")`, for example.  Though that wouldn't really be reliable because the key names for gamepad buttons change between DirectInput & XInput controllers.  The main concern would be making sure that all keys are named the same in the Windows & Linux versions (which is probably not the case for some keys).

Also, since you're using Lua scripts, you might want to try using this build: https://www.mesen.ca/MesenDebuggerFixes3.zip - it fixes a number of debugger issues that were recently reported to me (some of which are Lua-related)

**Comment by SourMesen** [comment]
This is what I'm planning at the moment for the save state issue:
`emu.saveSavestateAsync(function)`
`emu.loadSavestateAsync(string - binary)`

And then you could use it like this:
```
savestatedata = ""
function savedata(data)
  savestatedata = data;
end
emu.saveSavestateAsync(savedata);
```
The function passed to saveSavestateAsync will be called a single time, right at the start of the following CPU instruction.  Once you get the data, you can load it at any time by calling loadSavestateAsync(data) - the state will be loaded once the current instruction ends.

While I'm at it, I also want to add `stateSaved` / `stateLoaded` events to use with `addEventCallback`.

Would this solution work for you?

**Comment by gamax92** [comment]
Depends, what would be the range of slot numbers?

**Comment by SourMesen** [comment]
Any int would be simple enough to implement with a map - no real reason to limit it.
I like the slot numbers better too - the callback method is a bit annoying to use.

I should be able to add a special case for the cpuExec callback and make the async calls synchronous within that event.

**Comment by SourMesen** [comment]
I've added the following:
`emu.saveSavestateAsync(int)`
`emu.loadSavestateAsync(int)`
`emu.getSavestateData()`
`emu.clearSavestateData(int)`

save/load will trigger a save/load at the start of the next instruction.  If used within a cpuExec callback, the save/load operation will be immediate.  I think I'll also change the saveStatestate/loadSavestate functions to work within the cpuExec callbacks (but haven't yet)

I also added a `emu.isKeyPressed(string)` function - the key names match the 3rd value on each row in [this file](https://github.com/SourMesen/Mesen/blob/master/Windows/WindowsKeyManager.cpp) (so not the VK_.. column).

There's a build with these changes [here](https://www.mesen.ca/MesenDebuggerFixes4.zip) if you want to try it out.

**Comment by gamax92** [comment]
Thank you for the work, I'll try it out and report back.

**Comment by SourMesen** [comment]
I did some bug fixes for the Lua savestate functions (some of the load code, in particular, was wrong in some cases & could cause the game to crash after loading, etc).
saveSavestate & loadSavestate are now usable synchronously within any CpuExec callback, too (still the same api as the docs).
I added eventType.stateLoaded & eventType.stateSaved.  They are only triggered when the user manually loads or saves a state, anything else that loads/saves states (e.g the rewinder) does not trigger it.

Also, I changed some of the API slightly (debugRead/debugWrite have been replaced by memType.cpuDebug, memType.ppuDebug in the regular read/write functions instead) - just letting you know in case you were using them and end up wondering why they don't work anymore.

An updated build with these changes/fixes is [here](https://www.mesen.ca/MesenDebuggerFixes5.zip)
